### ISSUE TITLE
How to add metadata on collection?

### ISSUE THREAD

**Issue by mstfdkmn** [irods/irods_client_http_api#170]
### iRODS Version, OS and Version
4.3.1, almalinux8

### What did you try to do?
I tried to add metadata on a collection.

### Expected behavior
I expected to see metadata is attached.

### Observed behavior (including steps to reproduce, if applicable)
I tried with different languages but always got errors. I hope I am not making an error somewhere. Could you share with me an example that will show how you use curl to add atomic metadata?

Is [this ](https://github.com/irods/irods_client_http_api/blob/main/test/test_irods_http_api.py#L314-L330) passing?

**Comment by korydraughn** [comment]
All tests pass for 0.1.0.

Here's an example demonstrating how to add metadata to a collection using curl.
```bash
curl -H "Authorization: Bearer $token" 'http://localhost:9000/irods-http-api/0.1.0/collections' \
    --data-urlencode 'op=modify_metadata' \
    --data-urlencode 'lpath=/tempZone/home/alice' \
    --data-urlencode 'operations=[{"operation":"add", "attribute":"http-api", "value":"didit"}]'
```

On success, you can expect the following.
```json
{
  "irods_response": {
    "failed_operation": {},
    "status_code": 0
  }
}
```

**Comment by mstfdkmn** [comment]
I used the same example that you shared, but still getting a bad request error.

```
[irods@dev-u0137480 go-http-client]$ curl -H "Authorization: Bearer $token" 'http://localhost:9000/irods-http-api/0.1.0/collections' \
>     --data-urlencode 'op=modify_metadata' \
>     --data-urlencode 'lpath=/u0137480/home/rods' \
>     --data-urlencode 'operations=[{"operation":"add", "attribute":"http-api", "value":"didit"}]'
{"irods_response":{"status_code":-1}}
```

**Comment by korydraughn** [comment]
Yep. That definitely looks like a policy issue.

**Comment by korydraughn** [comment]
@mstfdkmn Any updates on this?

**Comment by mstfdkmn** [comment]
I may test this later. But if needed you may please close now.

**Comment by trel** [comment]
Let's leave it for you to determine things are working before we close.

**Comment by mstfdkmn** [comment]
Hi, I guess this is not a policy issue, it looks like related to `pep_api_atomic_apply_metadata_operations_*`. 

To reproduce:

- add the following in core.py
```
import json

def pep_api_atomic_apply_metadata_operations_pre(rule_args, callback, rei):
    """
    Triggered by PRC apply_atomic_operations() method
    Blocks rodsusers' remove requests of administrative metadata
    """
    obj_var = bytes(rule_args[2].get_bytes()).decode("utf8")
    callback.writeLine("serverLog", f"{obj_var}")
    json_var = json.loads(obj_var)
    callback.writeLine("serverLog", f"{json_var}")
```
- run the following to add metadata
`curl -H "Authorization: Bearer $token" 'http://localhost:9000/irods-http-api/0.1.0/collections'     --data-urlencode 'op=modify_metadata'     --data-urlencode 'lpath=/u0137480/home/rods'     --data-urlencode 'operations=[{"operation":"add", "attribute":"http-api", "value":"didit"}]'`

- in logs
```
irods stdout | {"log_category":"legacy","log_level":"info","log_message":"writeLine: inString = {\"admin_mode\":false,\"entity_name\":\"/u0137480/home/rods\",\"entity_type\":\"collection\",\"operations\":[{\"attribute\":\"http-api\",\"operation\":\"add\",\"value\":\"didit\"}]}\n","request_api_name":"","request_api_number":20002,"request_api_version":"d","request_client_user":"rods","request_host":"10.0.2.100","request_proxy_user":"rods","request_release_version":"rods4.3.1","server_host":"dev-u0137480.irods.t.icts.kuleuven.be","server_pid":3325801,"server_timestamp":"2024-03-15T13:44:23.170Z","server_type":"agent","server_zone":"u0137480"}
irods stdout | {"log_category":"rule_engine","log_level":"error","log_message":"caught python exception","python_exception":"Traceback (most recent call last):\n  File \"/etc/irods/core.py\", line 14, in pep_api_atomic_apply_metadata_operations_pre\n    json_var = json.loads(obj_var)\n  File \"/usr/lib64/python3.6/json/__init__.py\", line 354, in loads\n    return _default_decoder.decode(s)\n  File \"/usr/lib64/python3.6/json/decoder.py\", line 342, in decode\n    raise JSONDecodeError(\"Extra data\", s, end)\njson.decoder.JSONDecodeError: Extra data: line 1 column 158 (char 157)\n","request_api_name":"","request_api_number":20002,"request_api_version":"d","request_client_user":"rods","request_host":"10.0.2.100","request_proxy_user":"rods","request_release_version":"rods4.3.1","rule_engine_plugin":"python","server_host":"dev-u0137480.irods.t.icts.kuleuven.be","server_pid":3325801,"server_timestamp":"2024-03-15T13:44:23.170Z","server_type":"agent","server_zone":"u0137480"}
```

The solution provided at https://github.com/irods/irods/issues/6892 works for the PRC.

**Comment by korydraughn** [comment]
Sounds like this can be closed.

@mstfdkmn Do you agree?

**Comment by mstfdkmn** [comment]
I guess NO! When that pep like above is declared in core.py, there is no issue with the atomic metadata operations of the PRC. However when there is a call through the http-api, I encounter the issue mentioned. If you think the issue is with the `pep_api_atomic_apply_metadata_operations_pre`, should not it occur when fired by the PRC too?

**Comment by trel** [comment]
We'll try to reproduce what you have above.

**Comment by mstfdkmn** [comment]
We integrated the http-api to our production tier. This means that metadata operations that users will make/execute via the http api would fail in case there is a rule - `pep_api_atomic_apply_metadata_operations_*`. Would it be possible for you to test this (put a said pep rule in core.py and try modify_metadata operation)? Thanks.

**Comment by korydraughn** [comment]
We'll try to reproduce what you've described.

**Comment by korydraughn** [comment]
I've reproduced this. We'll get a fix in for the next release.

**Comment by korydraughn** [comment]
The call to `json.loads` appears to be tripped up by the last element within the byte sequence (i.e. a null byte).

Using `json.loads(obj_var[:-1])` resolves the issue, but only partially.

The reason the policy works with the PRC most likely has to do with the exclusion of the null byte. The C/C++ code includes it to help the server know where the string ends. See the following:
https://github.com/irods/irods/blob/e4b196d429c65c8d6333f524a0cdc33b4a870274/lib/api/src/rc_atomic_apply_metadata_operations.cpp#L18

The inclusion of the null byte isn't necessary because byte buffers include a member for defining the size of the byte sequence (i.e. `.len`). I'll open an issue for this.

For now, you can adjust your policy to accommodate C, C++, and Python by doing the following:
```python
import json

def pep_api_atomic_apply_metadata_operations_pre(rule_args, callback, rei):
    """
    Triggered by PRC apply_atomic_operations() method
    Blocks rodsusers' remove requests of administrative metadata
    """
    obj_var = bytes(rule_args[2].get_bytes()).decode("utf8")
    callback.writeLine("serverLog", f"{obj_var}")

    # Change how the JSON string is handled based on whether it ends with
    # a zero byte.
    bbuf = rule_args[2]
    if bbuf.get_byte(bbuf.len - 1) == 0:
        json_var = json.loads(obj_var[:-1])
    else:
        json_var = json.loads(obj_var)
    callback.writeLine("serverLog", f"{json_var}")
```

**Comment by mstfdkmn** [comment]
Thanks Kory. Now it works.


> The reason the policy works with the PRC most likely has to do with the exclusion of the null byte.

Fyi: I sometimes rarely encountered exact the same issue with the PRC. But I was not sure what was going on. And I could not always reproduce easily (could not sort it out in which condition). After I run the script that encountered the same issue second time, it was resolving. That is why I didn't emphasize this much with the PRC after the solution introduced [here](https://github.com/irods/irods/issues/6892#issuecomment-1419755415) and its root in the google chat.

**Comment by mstfdkmn** [comment]
> I'll open an issue for this.

Btw, for your awareness I have not seen an issue opened yet.

**Comment by korydraughn** [comment]
Created irods/irods/issues/7879.

**Comment by korydraughn** [comment]
Closing since the real issue must be resolved in the iRODS server.

**Comment by mstfdkmn** [comment]
> The call to `json.loads` appears to be tripped up by the last element within the byte sequence (i.e. a null byte).
> 
> Using `json.loads(obj_var[:-1])` resolves the issue, but only partially.
> 
> The reason the policy works with the PRC most likely has to do with the exclusion of the null byte. The C/C++ code includes it to help the server know where the string ends. See the following: https://github.com/irods/irods/blob/e4b196d429c65c8d6333f524a0cdc33b4a870274/lib/api/src/rc_atomic_apply_metadata_operations.cpp#L18
> 
> The inclusion of the null byte isn't necessary because byte buffers include a member for defining the size of the byte sequence (i.e. `.len`). I'll open an issue for this.
> 
> For now, you can adjust your policy to accommodate C, C++, and Python by doing the following:
> 
> ```python
> import json
> 
> def pep_api_atomic_apply_metadata_operations_pre(rule_args, callback, rei):
>     obj_var = bytes(rule_args[2].get_bytes()).decode("utf8")
>     callback.writeLine("serverLog", f"{obj_var}")
> 
>     # Change how the JSON string is handled based on whether it ends with
>     # a zero byte.
>     bbuf = rule_args[2]
>     if bbuf.get_byte(bbuf.len - 1) == 0:
>         json_var = json.loads(obj_var[:-1])
>     else:
>         json_var = json.loads(obj_var)
>     callback.writeLine("serverLog", f"{json_var}")
> ```
> 
> _Keep in mind the HTTP API needs to be restarted anytime the policy changes. A restart is only necessary if you are using connection pooling. If you run the HTTP API in 4.2 compatibility mode, then the changes to policy will become visible on each HTTP request._

Hello Kory, sorry but I am confused with the result of this python code when it is invoked by the PRC. Could you please try whether you can see the same result that I explained below?

Despite the fact that the relevant pep is fired only once, `callback.writeLine("serverLog", f"{json_var}")` sends json_var to the irods logs two times.

**Comment by korydraughn** [comment]
Sure. Please post the PRC code you are using to test it.

Just so it's clear my head, the rule code in question isn't failing, correct?

**Comment by mstfdkmn** [comment]
Okay, here it is:

```
with iRODSSession(irods_env_file=env_file, **ssl_settings) as session:
    avus = [('cold', 'true')]
    col = session.collections.get("/u0137480/home/mustafa/test_col")
    col.metadata.apply_atomic_operations(*[AVUOperation(operation='add', avu=iRODSMeta("{}".format(str(meta[0])), "{}".format(str(meta[1])))) for meta in avus])
```

**Comment by trel** [comment]
I was hoping that was the answer.

Consider adding more words to those two `callback.writeLine` lines to distinguish them from one another.

**Comment by mstfdkmn** [comment]
Surprising! I had already removed the `callback.writeLine("serverLog", f"{obj_var}")` but I am still getting two messages in the output as seen below.

```
@pep
def pep_api_atomic_apply_metadata_operations_post(rule_args, callback, rei):
    obj_var = bytes(rule_args[2].get_bytes()).decode("utf8")
    bbuf = rule_args[2]
    if bbuf.get_byte(bbuf.len - 1) == 0:
        json_var = json.loads(obj_var[:-1])
    else:
        json_var = json.loads(obj_var)
    callback.writeLine("serverLog", f"{json_var}")
    callback.writeLine("serverLog", "pep_api_atomic_apply_metadata_operations_post invoked!")
```

irods logs:
```
{"processname": "irods", "groupname": "irods", "pid": "125", "channel": "stdout", "irods": {"log_category": "legacy", "log_level": "info", "log_message": "writeLine: inString = {'entity_name': '/u0137480/home/mustafa/test_col', 'entity_type': 'collection', 'operations': [{'operation': 'add', 'attribute': 'cold', 'value': 'true'}]}\n", "request_api_name": "", "request_api_number": 20002, "request_api_version": "d", "request_client_user": "mustafa", "request_host": "10.0.2.100", "request_proxy_user": "mustafa", "request_release_version": "rods4.3.2", "server_host": "dev-u0137480.irods.t.icts.kuleuven.be", "server_pid": 187398, "server_timestamp": "2024-07-12T14:30:38.046Z", "server_type": "agent", "server_zone": "u0137480"}}
{"processname": "irods", "groupname": "irods", "pid": "125", "channel": "stdout", "irods": {"log_category": "legacy", "log_level": "info", "log_message": "writeLine: inString = {'entity_name': '/u0137480/home/mustafa/test_col', 'entity_type': 'collection', 'operations': [{'operation': 'add', 'attribute': 'cold', 'value': 'true'}]}\n", "request_api_name": "", "request_api_number": 20002, "request_api_version": "d", "request_client_user": "mustafa", "request_host": "10.0.2.100", "request_proxy_user": "mustafa", "request_release_version": "rods4.3.2", "server_host": "dev-u0137480.irods.t.icts.kuleuven.be", "server_pid": 187398, "server_timestamp": "2024-07-12T14:30:38.094Z", "server_type": "agent", "server_zone": "u0137480"}}
{"processname": "irods", "groupname": "irods", "pid": "125", "channel": "stdout", "irods": {"log_category": "legacy", "log_level": "info", "log_message": "writeLine: inString = pep_api_atomic_apply_metadata_operations_post invoked!\n", "request_api_name": "", "request_api_number": 20002, "request_api_version": "d", "request_client_user": "mustafa", "request_host": "10.0.2.100", "request_proxy_user": "mustafa", "request_release_version": "rods4.3.2", "server_host": "dev-u0137480.irods.t.icts.kuleuven.be", "server_pid": 187398, "server_timestamp": "2024-07-12T14:30:38.094Z", "server_type": "agent", "server_zone": "u0137480"}}
```

**Comment by mstfdkmn** [comment]
more info:

```
def pep_api_atomic_apply_metadata_operations_post(rule_args, callback, rei):
    obj_var = bytes(rule_args[2].get_bytes()).decode("utf8")
    bbuf = rule_args[2]
    if bbuf.get_byte(bbuf.len - 1) == 0:
        json_var = json.loads(obj_var[:-1])
        callback.writeLine("serverLog", f"inside the if statement: {json_var}")
    else:
        json_var = json.loads(obj_var)
        callback.writeLine("serverLog", f"inside the else statement: {json_var}")

    callback.writeLine("serverLog", f"outside the if/else block: {json_var}")
    callback.writeLine("serverLog", "pep_api_atomic_apply_metadata_operations_post invoked!")
```

logs:
```
{"processname": "irods", "groupname": "irods", "pid": "125", "channel": "stdout", "irods": {"log_category": "legacy", "log_level": "info", "log_message": "writeLine: inString = {'entity_name': '/u0137480/home/mustafa/test_col', 'entity_type': 'collection', 'operations': [{'operation': 'add', 'attribute': 'cold', 'value': 'true'}]}\n", "request_api_name": "", "request_api_number": 20002, "request_api_version": "d", "request_client_user": "mustafa", "request_host": "10.0.2.100", "request_proxy_user": "mustafa", "request_release_version": "rods4.3.2", "server_host": "dev-u0137480.irods.t.icts.kuleuven.be", "server_pid": 187491, "server_timestamp": "2024-07-12T14:36:26.584Z", "server_type": "agent", "server_zone": "u0137480"}}
{"processname": "irods", "groupname": "irods", "pid": "125", "channel": "stdout", "irods": {"log_category": "legacy", "log_level": "info", "log_message": "writeLine: inString = inside the else statement: {'entity_name': '/u0137480/home/mustafa/test_col', 'entity_type': 'collection', 'operations': [{'operation': 'add', 'attribute': 'cold', 'value': 'true'}]}\n", "request_api_name": "", "request_api_number": 20002, "request_api_version": "d", "request_client_user": "mustafa", "request_host": "10.0.2.100", "request_proxy_user": "mustafa", "request_release_version": "rods4.3.2", "server_host": "dev-u0137480.irods.t.icts.kuleuven.be", "server_pid": 187491, "server_timestamp": "2024-07-12T14:36:26.623Z", "server_type": "agent", "server_zone": "u0137480"}}
{"processname": "irods", "groupname": "irods", "pid": "125", "channel": "stdout", "irods": {"log_category": "legacy", "log_level": "info", "log_message": "writeLine: inString = outside the if/else block: {'entity_name': '/u0137480/home/mustafa/test_col', 'entity_type': 'collection', 'operations': [{'operation': 'add', 'attribute': 'cold', 'value': 'true'}]}\n", "request_api_name": "", "request_api_number": 20002, "request_api_version": "d", "request_client_user": "mustafa", "request_host": "10.0.2.100", "request_proxy_user": "mustafa", "request_release_version": "rods4.3.2", "server_host": "dev-u0137480.irods.t.icts.kuleuven.be", "server_pid": 187491, "server_timestamp": "2024-07-12T14:36:26.624Z", "server_type": "agent", "server_zone": "u0137480"}}
{"processname": "irods", "groupname": "irods", "pid": "125", "channel": "stdout", "irods": {"log_category": "legacy", "log_level": "info", "log_message": "writeLine: inString = pep_api_atomic_apply_metadata_operations_post invoked!\n", "request_api_name": "", "request_api_number": 20002, "request_api_version": "d", "request_client_user": "mustafa", "request_host": "10.0.2.100", "request_proxy_user": "mustafa", "request_release_version": "rods4.3.2", "server_host": "dev-u0137480.irods.t.icts.kuleuven.be", "server_pid": 187491, "server_timestamp": "2024-07-12T14:36:26.624Z", "server_type": "agent", "server_zone": "u0137480"}}
```

**Comment by korydraughn** [comment]
And restarting the server results in the same behavior?

**Comment by mstfdkmn** [comment]
We never restarted in the past the irods server in order for the PRC connection to see policy changes in the server. I guess the PRC should immediately see

**Comment by trel** [comment]
So your last output section appears to be behaving as expected?

**Comment by mstfdkmn** [comment]
I guess yes, these `callback.writeLine("serverLog", f"outside the if/else block: {json_var}")`, `callback.writeLine("serverLog", "pep_api_atomic_apply_metadata_operations_post invoked!")` looks okay. The second line in the output looks okay too.

But I am still confused with the first line in the output. How that is possible?

**Comment by korydraughn** [comment]
Seems like something is being cached, which is why I was asking if the behavior changes following a restart.

And yes, you shouldn't need to restart the server for changing policy, but it's unclear how that's happening at the moment.

**Comment by korydraughn** [comment]
That last output does appear to be working now, somehow.

**Comment by mstfdkmn** [comment]
> Seems like something is being cached, which is why I was asking if the behavior changes following a restart.

I restarted the irods server and I am still seeing the same result.

**Comment by korydraughn** [comment]
Is there any chance that you have multiple PEPs in use for the atomic metadata API (i.e. `_pre` and `_post`)?

**Comment by korydraughn** [comment]
I noticed references to `_pre` and `_post` for the atomic PEPs in this thread, which is why I asked about multiple PEPs.

**Comment by mstfdkmn** [comment]
> Is there any chance that you have multiple PEPs in use for the atomic metadata API (i.e. `_pre` and `_post`)?

No, there was only one.

But okay the issue looks like indeed related to something is being cached...Because when I try this in a fresh irods container (so a new irods with a new container), I see everything again works normal.

**Comment by korydraughn** [comment]
Yeah, very weird. If you run into anymore issues like this, please open a new issue.
